### ISSUE TITLE
Implement in TypeScript

### ISSUE THREAD

**Issue by jimmy4701** [altangent/ccxws#229]
The project is perfect but we'd need a typescript implementation of ccxws, to work well in a Typescript environment.

**Maybe just a declaration file**
Maybe just a declaration file (like in ccxt) would do the job, instead of re-developping all the project in TS ?

**Comment by bmancini55** [comment]
@jimmy4701 thanks for submitting the issue!

We're planning to fully convert the library to TypeScript with the refactor in #149. 

Until that happens, a declaration file certainly seems like a easy enough path forward.  Happy to accept a PR for it if there is interest in helping with the library!

**Comment by jimmy4701** [comment]
@bmancini55 , thanks for your reply .

No problem, so we'll wait until the push of #149 👍 

We're already using a declaration file, but a complete implementation of the types would be more effective :P 
I also already asked the dev team to develop it by themself, to make our first PR to CCXWS, cause it would be very cool to help the project, after the lot of time it made us winning :D So it would be our pleasure to help you.

**Comment by Shepless** [comment]
@jimmy4701 If you need type definitions until the migration is done:

1) Create a `typings` directory
2) Create a new file `ccxws.d.ts` file
3) Copy and paste type definitions (below)
4) Update your `tsconfig.json` to include the new typings (example below)

### tsconfig.json
```js
"compilerOptions": {
        // ... your existing config
        "typeRoots": [
            "./node_modules/@types",
            "./typings"
        ]
}
```

### Type Defintions
```ts
declare module "ccxws" {
    type ExchangeCtor = new (options?: IExchangeOptions) => IExchange;
    type ConnectionEventType = 'connecting' | 'connected' | 'disconnected' | 'closing' | 'closed' | 'reconnecting';
    type ExchangeEventType = 'ticker' | 'candle' | 'trade' | 'l2snapshot' | 'l2update' | 'l3snapshot' | 'l3update' | 'error';
    type EventType = ConnectionEventType | ExchangeEventType;
    type CandlePeriod = {
        _1m: '_1m',
        _2m: '_2m',
        _3m: '_3m',
        _5m: '_5m',
        _15m: '_15m',
        _30m: '_30m',
        _1h: '_1h',
        _2h: '_2h',
        _4h: '_4h',
        _6h: '_6h',
        _8h: '_8h',
        _12h: '_12h',
        _1d: '_1d',
        _3d: '_3d',
        _1w: '_1w',
        _2w: '_2w',
        _1M: '_1M'
    };

    interface ITicker {
        exchange: string;
        base: string;
        quote: string;
        timestamp: number;
        last: string;
        open: string;
        low: string;
        high: string;
        volume: string;
        quoteVolume: string;
        change: string;
        changePercent: string;
        bid: string;
        bidVolume: string;
        ask: string;
        askVolume: string;
    }

    interface ITrade {
        exchange: string;
        base: string;
        quote: string;
        tradeId: string;
        unix: number;
        side: string;
        price: string;
        amount: string;
        buyOrderId?: string;
        sellOrderId?: string;
    }

    interface ICandle {
        timestampMs: number;
        open: string;
        high: string;
        low: string;
        close: string;
        volume: string;
    }

    interface ILevelData<T> {
        exchange: string;
        base: string;
        quote: string;
        timestampMs?: number;
        sequenceId?: number;
        asks: T[];
        bids: T[];
    }

    interface ILevel2Point {
        price: string;
        size: string;
        count?: number;
    }

    interface ILevel3Point {
        orderId: string;
        price: string;
        size: string;
        meta?: any;
    }

    interface ILevel2Data extends ILevelData<ILevel2Point> { }

    interface ILevel3Data extends ILevelData<ILevel3Point> { }

    interface IMarket {
        id: string;
        base: string;
        quote: string;
        type?: string;
    }

    interface IExchangeOptions {
        wssPath?: string;
        watcherMs?: number;
        apiKey?: string
        apiSecret?: string;
    }

    interface EventCallbackMap {
        'connecting': () => void;
        'connected': () => void;
        'disconnected': () => void;
        'closing': () => void;
        'closed': () => void;
        'reconnecting': () => void;
        'error': (error: Error) => void;
        'ticker': (ticker: ITicker, market: IMarket) => void;
        'trade': (trade: ITrade, market: IMarket) => void;
        'candle': (candle: ICandle, market: IMarket) => void;
        'l2snapshot': (snapshot: ILevel2Data, market: IMarket) => void;
        'l2update': (snapshot: ILevel2Data, market: IMarket) => void;
        'l3snapshot': (snapshot: ILevel3Data, market: IMarket) => void;
        'l3update': (snapshot: ILevel3Data, market: IMarket) => void;
    }

    interface IExchange {
        candlePeriod: keyof CandlePeriod;
        hasTickers: boolean;
        hasTrades: boolean;
        hasCandles: boolean;
        hasLevel2Snapshots: boolean;
        hasLevel2Updates: boolean;
        hasLevel3Snapshots: boolean;
        hasLevel3Updates: boolean;
        on<T extends EventType>(event: T, callback: EventCallbackMap[T]): void;
        subscribeTicker(market: IMarket): void;
        unsubscribeTicker(market: IMarket): void;
        subscribeTrades(market: IMarket): void;
        unsubscribeTrades(market: IMarket): void;
        subscribeCandles(market: IMarket): void;
        unsubscribeCandles(market: IMarket): void;
        subscribeLevel2Snapshots(market: IMarket): void;
        unsubscribeLevel2Snapshots(market: IMarket): void;
        subscribeLevel2Updates(market: IMarket): void;
        unsubscribeLevel2Updates(market: IMarket): void;
        subscribeLevel3Snapshots(market: IMarket): void;
        unsubscribeLevel3Snapshots(market: IMarket): void;
        subscribeLevel3Updates(market: IMarket): void;
        unsubscribeLevel3Updates(market: IMarket): void;
    }

    interface ICCXWS {
        Bibox: ExchangeCtor;
        Binance: ExchangeCtor;
        BinanceFuturesCoinM: ExchangeCtor;
        BinanceFuturesUsdtM: ExchangeCtor;
        BinanceJe: ExchangeCtor;
        BinanceUs: ExchangeCtor;
        Bitfinex: ExchangeCtor;
        Bitflyer: ExchangeCtor;
        Bithumb: ExchangeCtor;
        BitMEX: ExchangeCtor;
        Bitstamp: ExchangeCtor;
        Bittrex: ExchangeCtor;
        Cex: ExchangeCtor;
        CoinbasePro: ExchangeCtor;
        Coinex: ExchangeCtor;
        Deribit: ExchangeCtor;
        Digifinex: ExchangeCtor;
        Ethfinex: ExchangeCtor;
        ErisX: ExchangeCtor;
        Ftx: ExchangeCtor;
        FtxUs: ExchangeCtor;
        Gateio: ExchangeCtor;
        Gemini: ExchangeCtor;
        HitBTC: ExchangeCtor;
        Huobi: ExchangeCtor;
        HuobiFutures: ExchangeCtor;
        HuobiSwaps: ExchangeCtor;
        HuobiJapan: ExchangeCtor;
        HuobiKorea: ExchangeCtor;
        HuobiRussia: ExchangeCtor;
        Kucoin: ExchangeCtor;
        Kraken: ExchangeCtor;
        LedgerX: ExchangeCtor;
        Liquid: ExchangeCtor;
        OKEx: ExchangeCtor;
        Poloniex: ExchangeCtor;
        Upbit: ExchangeCtor;
        Zb: ExchangeCtor;
        CandlePeriod: CandlePeriod;
    }

    const ccxws: ICCXWS;
    export default ccxws;
}
```

**Comment by bmancini55** [comment]
@Shepless thanks for putting together typedefs!  If you put those into a PR I will gladly merge it!

**Comment by Industrial** [comment]
Hi!

I would like to make these edits:

1. Remove 'I' from all interface names. This is an old convention and eslint recommends against it.
2. Change all interfaces to types. Having coded TypeScript for years now, I don't see the need for interfaces at all. You can do everything you want with Type Intersections and Type Unions.
3. Export all types.

```
declare module 'ccxws' {
  type ExchangeCtor = new (options?: ExchangeOptions) => Exchange
  type ConnectionEventType = 'connecting' | 'connected' | 'disconnected' | 'closing' | 'closed' | 'reconnecting'
  type ExchangeEventType =
    | 'ticker'
    | 'candle'
    | 'trade'
    | 'l2snapshot'
    | 'l2update'
    | 'l3snapshot'
    | 'l3update'
    | 'error'
  type EventType = ConnectionEventType | ExchangeEventType
  type CandlePeriod = {
    _1m: '_1m'
    _2m: '_2m'
    _3m: '_3m'
    _5m: '_5m'
    _15m: '_15m'
    _30m: '_30m'
    _1h: '_1h'
    _2h: '_2h'
    _4h: '_4h'
    _6h: '_6h'
    _8h: '_8h'
    _12h: '_12h'
    _1d: '_1d'
    _3d: '_3d'
    _1w: '_1w'
    _2w: '_2w'
    _1M: '_1M'
  }

  export type Ticker = {
    exchange: string
    base: string
    quote: string
    timestamp: number
    last: string
    open: string
    low: string
    high: string
    volume: string
    quoteVolume: string
    change: string
    changePercent: string
    bid: string
    bidVolume: string
    ask: string
    askVolume: string
  }

  export type Trade = {
    exchange: string
    base: string
    quote: string
    tradeId: string
    unix: number
    side: string
    price: string
    amount: string
    buyOrderId?: string
    sellOrderId?: string
  }

  export type Candle = {
    timestampMs: number
    open: string
    high: string
    low: string
    close: string
    volume: string
  }

  export type LevelData<T> = {
    exchange: string
    base: string
    quote: string
    timestampMs?: number
    sequenceId?: number
    asks: T[]
    bids: T[]
  }

  export type Level2Point = {
    price: string
    size: string
    count?: number
  }

  export type Level3Point = {
    orderId: string
    price: string
    size: string
    meta?: any
  }

  export type Level2Data = LevelData<Level2Point> & {}

  export type Level3Data = LevelData<Level3Point> & {}

  export type Market = {
    id: string
    base: string
    quote: string
    type?: string
  }

  export type ExchangeOptions = {
    wssPath?: string
    watcherMs?: number
    apiKey?: string
    apiSecret?: string
  }

  export type EventCallbackMap = {
    connecting: () => void
    connected: () => void
    disconnected: () => void
    closing: () => void
    closed: () => void
    reconnecting: () => void
    error: (error: Error) => void
    ticker: (ticker: Ticker, market: Market) => void
    trade: (trade: Trade, market: Market) => void
    candle: (candle: Candle, market: Market) => void
    l2snapshot: (snapshot: Level2Data, market: Market) => void
    l2update: (snapshot: Level2Data, market: Market) => void
    l3snapshot: (snapshot: Level3Data, market: Market) => void
    l3update: (snapshot: Level3Data, market: Market) => void
  }

  export type Exchange = {
    candlePeriod: keyof CandlePeriod
    hasTickers: boolean
    hasTrades: boolean
    hasCandles: boolean
    hasLevel2Snapshots: boolean
    hasLevel2Updates: boolean
    hasLevel3Snapshots: boolean
    hasLevel3Updates: boolean
    on<T extends EventType>(event: T, callback: EventCallbackMap[T]): void
    subscribeTicker(market: Market): void
    unsubscribeTicker(market: Market): void
    subscribeTrades(market: Market): void
    unsubscribeTrades(market: Market): void
    subscribeCandles(market: Market): void
    unsubscribeCandles(market: Market): void
    subscribeLevel2Snapshots(market: Market): void
    unsubscribeLevel2Snapshots(market: Market): void
    subscribeLevel2Updates(market: Market): void
    unsubscribeLevel2Updates(market: Market): void
    subscribeLevel3Snapshots(market: Market): void
    unsubscribeLevel3Snapshots(market: Market): void
    subscribeLevel3Updates(market: Market): void
    unsubscribeLevel3Updates(market: Market): void
  }

  export type CCXWS = {
    Bibox: ExchangeCtor
    Binance: ExchangeCtor
    BinanceFuturesCoinM: ExchangeCtor
    BinanceFuturesUsdtM: ExchangeCtor
    BinanceJe: ExchangeCtor
    BinanceUs: ExchangeCtor
    Bitfinex: ExchangeCtor
    Bitflyer: ExchangeCtor
    Bithumb: ExchangeCtor
    BitMEX: ExchangeCtor
    Bitstamp: ExchangeCtor
    Bittrex: ExchangeCtor
    Cex: ExchangeCtor
    CoinbasePro: ExchangeCtor
    Coinex: ExchangeCtor
    Deribit: ExchangeCtor
    Digifinex: ExchangeCtor
    Ethfinex: ExchangeCtor
    ErisX: ExchangeCtor
    Ftx: ExchangeCtor
    FtxUs: ExchangeCtor
    Gateio: ExchangeCtor
    Gemini: ExchangeCtor
    HitBTC: ExchangeCtor
    Huobi: ExchangeCtor
    HuobiFutures: ExchangeCtor
    HuobiSwaps: ExchangeCtor
    HuobiJapan: ExchangeCtor
    HuobiKorea: ExchangeCtor
    HuobiRussia: ExchangeCtor
    Kucoin: ExchangeCtor
    Kraken: ExchangeCtor
    LedgerX: ExchangeCtor
    Liquid: ExchangeCtor
    OKEx: ExchangeCtor
    Poloniex: ExchangeCtor
    Upbit: ExchangeCtor
    Zb: ExchangeCtor
    CandlePeriod: CandlePeriod
  }

  const ccxws: CCXWS
  export default ccxws
}
```

**Comment by Shepless** [comment]
You're welcome 👍

**Comment by bmancini55** [comment]
Closed with #292
